csrc/preseg_passes/make_resharding_contiguous.cpp

```diff
@@ -12,35 +12,139 @@
 
 namespace nvfuser::preseg_passes {
 
 namespace {
-void setShardedAllocationDomain(TensorView* tv) {
-  if (!tv->hasAllocation()) {
-    tv->setAllocationDomain(tv->getLoopDomain(), true);
+
+// Validates meshes (i.e. all TensorViews have a device mesh or none) and
+// returns true if any TensorView has a device mesh.
+bool validateMeshes(Fusion* fusion) {
+  // Validate that meshes are assigned to all TensorViews or none.
+  bool tv_with_mesh_found = false;
+  bool tv_without_mesh_found = false;
+
+  for (auto tv : fusion->allTvs()) {
+    if (tv->isCpuScalar()) {
+      continue;
+    }
+    tv->hasDeviceMesh() ? tv_with_mesh_found = true
+                        : tv_without_mesh_found = true;
+  }
+  NVF_CHECK(
+      !(tv_with_mesh_found && tv_without_mesh_found),
+      "Cannot have some TensorViews with device mesh and some without.");
+  return tv_with_mesh_found;
+}
+
+// Reorders the loop domain in the same relative order as the allocation domain.
+// Specifically:
+// 1. It uses the exprs between logical and loop domain to split the allocation
+// domain
+// 2. It reorders the loop domain to match the split allocation domain.
+// 3. It computes the contiguity of the transformed allocation domain through
+// the split exprs.
+// 4. Sets the allocation domain to be the same as the loop domain with the
+// computed contiguity. This preserves both the sharding and any stride order.
+// Note: Ideally, the loop domain can follow the logical domain and the
+// allocation domain can follow the stride order specified/inferred. However, we
+// currently require loop domain to be the same as allocation domain. This
+// behavior will be modified in the future with allocation and loop domain being
+// propagated independently.
+void setLoopAndAllocationDomain(TensorView* tv) {
+  auto alloc_dom = tv->getMaybeAllocationDomain();
+  auto contiguity = tv->getContiguity();
+
+  auto splitContiguity = [](std::optional<bool> contiguity)
+      -> std::pair<std::optional<bool>, std::optional<bool>> {
+    if (!contiguity.has_value()) {
+      return std::make_pair(std::nullopt, std::nullopt);
+    }
+    if (contiguity.value()) {
+      return std::make_pair(true, true);
+    }
+    return std::make_pair(true, false);
+  };
+
+  // Allocation domain should be a permutation of logical domain at this point.
+  std::vector<Expr*> transform_exprs = DependencyCheck::getAllExprsBetween(
+      {alloc_dom.begin(), alloc_dom.end()},
+      {tv->getLoopDomain().begin(), tv->getLoopDomain().end()});
+
+  NVF_ERROR(
+      std::all_of(
+          transform_exprs.begin(),
+          transform_exprs.end(),
+          [](Expr* expr) { return expr->isA<Split>(); }),
+      "Expected all transform exprs to be a split between logical and loop domain during sharding propagation.");
+
+  for (auto* expr : transform_exprs) {
+    Split* split = dynamic_cast<Split*>(expr);
+    auto find_it = std::find(alloc_dom.begin(), alloc_dom.end(), split->in());
```

Comment:
FYI, you can implement this faster with https://github.com/NVIDIA/Fuser/blob/13a879c72bf304b6fb13cdd266bad066c71cc6d0/csrc/linked_hash_map.h#L4. Feel free to do that in a separate PR

Reply:
Listed it as a follow-up: https://github.com/NVIDIA/Fuser/issues/4298.